### PR TITLE
Refactoring, Add Helper, change convert method params, add rate method

### PR DESCRIPTION
**Pull request by JeanValJeann**
Hi @AlessandroMinoccheri ,

I purpose you my new version (3.1.0)

What I did is:

- refactored Component
- created an Helper
- changed reamde.md for a full explantion of how to use Plugin via /docs folder
- modified convert method parameters order to be more clear
- added rate method in component and helper to simply get a currency rate
- added a sql schema in config/table.txt to allow user to create table wich stores rates if they use database. I do not pass through Component/Helper to create table now, in this way we save one database request at each Component/Helper call.  How to set database use and create table is specified in docs.

Check what I am purposing to you! ;)

### CODE REVIEW

File: src/Controller/Component/CurrencyConverterComponent.php [hunk 1]
```diff
@@ -5,207 +5,174 @@
...
-            $now = date('Y-m-d H:i:s');
-            $dStart = new \DateTime($now);
-            $diff = $dStart->diff($lastUpdated);
-
-            if ($this->ensureNeedToUpdateDatabase($diff, $row)) {
-                $this->updateDatabase($row);
+        $rate = floatval(1);
+
+        if ($from !== $to) {
+
```
Comment (AlessandroMinoccheri) [inline src/Controller/Component/CurrencyConverterComponent.php]: please remove this empty lines

File: src/Controller/Component/CurrencyConverterComponent.php [hunk 2]
```diff
@@ -5,207 +5,174 @@
...
+                $result = $this->_currencyratesTable
+                    ->find('all')
+                    ->where([
+                        'from_currency' => $from,
+                        'to_currency' => $to
+                    ])
+                    ->first();
+
+                // If currency rate is in table and it doesn't have to be updated
+                if ($result && $result->modified->wasWithinLast($this->refresh . ' hours')) $rate = $result->rate;
```
Comment (AlessandroMinoccheri) [inline src/Controller/Component/CurrencyConverterComponent.php]: Please use psr standard in this if statement

File: src/Controller/Component/CurrencyConverterComponent.php [hunk 3]
```diff
@@ -5,207 +5,174 @@
...
+                            'from_currency' => $from,
+                            'to_currency' => $to,
+                            'rate' => $rate
+                        ]);
+                        $this->_currencyratesTable->save($entity);
+                    }
+                }
             } else {
-                $this->rate = $row['rates'];
+                $rate = $this->_getRateFromAPI($from, $to);
```
Comment (AlessandroMinoccheri) [inline src/Controller/Component/CurrencyConverterComponent.php]: I think that this else conditions could be removed

File: tests/bootstrap.php [hunk 4]
```diff
@@ -1,100 +1,84 @@
...
+define('APP', sys_get_temp_dir());
+define('APP_DIR', 'src');
+define('CAKE_CORE_INCLUDE_PATH', ROOT . '/vendor/cakephp/cakephp');
+define('CORE_PATH', CAKE_CORE_INCLUDE_PATH . DS);
+define('CAKE', CORE_PATH . APP_DIR . DS);
+define('WWW_ROOT', ROOT . DS . 'webroot' . DS);
+define('CONFIG', dirname(__FILE__) . DS . 'config' . DS);
 
-error_reporting(E_WARNING);
+ini_set('intl.default_locale', 'fr_FR');
```
Comment (AlessandroMinoccheri) [inline tests/bootstrap.php]: is necessary this configuration?

File: src/Controller/Component/CurrencyConverterComponent.php (hunk 1 above)
Comment (JeanValJeann) [reply src/Controller/Component/CurrencyConverterComponent.php]: Yes I do it

File: tests/bootstrap.php (hunk 4 above)
Comment (JeanValJeann) [reply tests/bootstrap.php]: No, it wasn't

**Comment by JeanValJeann** [comment]
@AlessandroMinoccheri,
I commited in my pull request what you wanted to be changed.

**Comment by AlessandroMinoccheri** [comment]
Now there is some conflicts and if is possible squash your commit @JeanValJeann

**Comment by JeanValJeann** [comment]
Hi, whats the conflict ?

**Comment by AlessandroMinoccheri** [comment]
Here in github I seen this conflict:

src/Controller/Component/CurrencyConverterComponent.php
tests/bootstrap.php

**Comment by JeanValJeann** [comment]
I saw it! It was 2 conflicts about currency's API request and bootstrap. Now, that's be ok?

**Comment by AlessandroMinoccheri** [comment]
ok

**Comment by JeanValJeann** [comment]
@AlessandroMinoccheri  commits has been squashed

**Comment by JeanValJeann** [comment]
hi @AlessandroMinoccheri ,
Is all right by your side?

**Comment by AlessandroMinoccheri** [comment]
I have already conflicts with this pull request into these files:

src/Controller/Component/CurrencyConverterComponent.php
tests/bootstrap.php

I have accepted it but now I can't merge without resolve conflicts

**Comment by JeanValJeann** [comment]
Ok @AlessandroMinoccheri !
Conflict has been resolved!
All Test passed!

Can you merge it now?
Should we probably create 3.1.0 tag to avoid having trouble for user who actually use plugin?

**Comment by JeanValJeann** [comment]
Nice that you have not merged this commit yet. I had time to add Session feature, now, Plugin use Session to avoid querying at each use case when using database. All test still passed!

**Comment by AlessandroMinoccheri** [comment]
Merged now. Sorry but yesterday was holiday and I have a family also. Thanks for your contribution! It's very important for me developer's that want to contribute to this project. Thanks a lot

**Pull request merged**